docs/proposals/L7-traffic-route.md

```diff
@@ -0,0 +1,174 @@
+            value: xyz
+            append: true
+        remove:
+          - name: x-something
+    destination: # required either split or a destination. Destination is a syntax sugar over split to one destination with weight 100 
+      kuma.io/service: usr_svc_6379
+      version: '1.0'
+    split:
+      - weight: 90
+        destination:
+          kuma.io/service: usr_svc_6379
+          version: '1.0'
+      - weight: 10
+        destination:
+          kuma.io/service: usr_svc_6379
+          version: '2.0'
+  destination: # if we have a http section and we don't match anything, then this rule is applied. It's again destination or split
+    kuma.io/service: usr_svc_6379
+    version: '1.0'
+  split:
+    - weight: 100
+      destination:
+        kuma.io/service: usr_svc_6379
+        version: '1.0'
+```
+
+### More examples
+
+1) Canary deployment
+
+**Use case:** when we release something on prod, but want to check functionality before serving it to all the users
+
+When any service is trying to consume `backend` with header `canary: "true"` the traffic is redirected to 1.1 backend, otherwise the traffic is passed to a 1.0 backend.
+
+```yaml
+type: TrafficRoute
+name: canary
+mesh: default
+sources:
+- match:
+    kuma.io/service: '*'
+destinations:
+- match:
+    kuma.io/service: backend
+conf:
+  http:
+  - match:
+      headers:
+        canary:
+          exact: "true"
+    destination:
+      kuma.io/service: backend
+      version: '1.1'
+  destination:
+    kuma.io/service: backend
+    version: '1.0'
+```
+
+2) Redirecting traffic to a different service
+
+**Use case:** Extracting a new `offers` microservice from `backend`, we can gradually shift the traffic with the `TrafficRoute`
+
+```yaml
+type: TrafficRoute
+name: reroute
+mesh: default
+sources:
+- match:
+    kuma.io/service: '*'
+destinations:
+- match:
+    kuma.io/service: backend
+conf:
+  http:
+  - match:
+      path:
+        prefix: "/offers"
+    destination:
+      kuma.io/service: offers
+  destination:
+    kuma.io/service: backend
+```
+
+3) Generic canary
+
+With the "traditional" `split` there is support for `*` in the tags. We should also support this use case in `http`.
+So this definition is similar to 1) but more generic.
+
+```yaml
+type: TrafficRoute
+name: canary
+mesh: default
+sources:
+- match:
+    kuma.io/service: '*'
+destinations:
+- match:
+    kuma.io/service: '*'
+conf:
+  http:
+  - match:
+      headers:
+        canary:
+          exact: "true"
+    split:
+      - destination:
+          kuma.io/service: '*'
+          version: canary
+  destination:
+    kuma.io/service: '*'
+```
+
+## Other notes
+
+* Split will require weight
+
+## Matching
+
+`TrafficRoute` will preserve same behavior of matching, meaning that we will only take into account `sources` and `destinations`
+for matching and then take `http` section from it.
+
+## Protocol
+
+If we have `http` section, it will be only applied on HTTP traffic.
```

Comment:
Assume that "HTTP" in this context means, http, https, http/2 and so on?

Reply:
`https` yes, but only when mTLS is provided by Kuma. Otherwise, we cannot decrypt the traffic

Reply:
Ok you confused me with mtls :) why is mtls required here?

Reply:
Imagine there is a client and server with sidecars
```
A <-> A's sidecar <-> B's sidecar <-> B
```
If A _itself_ does not encrypt the traffic then even with mTLS A's sidecar can redirect the traffic because it can read the request metadata (it happens before encrypting the traffic)

However, if A encrypts the traffic with TLS itself (for example old service when migrating to mesh), then A's sidecar cannot decrypt the traffic and we cannot route it by matching path, headers, etc.